### ISSUE TITLE
Save and load search results.

### ISSUE THREAD

**Issue by betolink** [nsidc/earthaccess#306]
There are use cases when we would like to store the results of a search for future use. One of such use cases is for when we are testing the same set of files or datasets, or in case CMR is down. This will look like:

```python
results = earthaccess.search_data(params)
earthaccess.save(results, "query.json")

# later on...

results = earthaccess.load("query.json")
```

The query will also have the original query to CMR to compare integrity.

**Comment by Sherwin-14** [comment]
Based on some conversations with @mfisher87, I concluded that the implementation of `Results` class will make it easier for us to add this functionality. We can add methods within the class itself that makes it easier to for us to encode this into JSON(and subsequently making use of those methods to create a `save` and a `load` function). I'm curious as to what others think of this issue? I am willing to help out in reviving [this](https://github.com/nsidc/earthaccess/pull/898) PR if it helps in solving this issue.

@chuckwondo @jhkennedy @betolink  Thoughts?

**Comment by chuckwondo** [comment]
I'm not convinced we need this. In the context of testing, we use vcrpy for recording requests and responses. I believe this issue was created before we started making use of vcrpy, so perhaps there's less need for this functionality now? Is there a strong desire for such functionality _outside_ of testing?

**Comment by mfisher87** [comment]
I could see saving and loading results being useful for reproducibility. CMR results can change, either due to mistakes, deprecation of datasets, or CMR going away entirely. Of course there's nothing we can do if a dataset is removed from its hosted location. But saving the CMR results and committing them would be useful context that would otherwise not be available to would-be reproducers in the future.

**Comment by chuckwondo** [comment]
In that case, perhaps we might consider how we might leverage [requests-cache](https://github.com/requests-cache/requests-cache) for a more general mechanism that won't require us to reinvent the wheel.

**Comment by jhkennedy** [comment]
My understanding of this issue is that it's not about caching the actual CMR responses, but being able to serialize the earthaccess search results (that is, the returned data structure). That can be used for testing, but it's also useful for a variety of purposes beyond that. Particularly, any time you're building out workflows on top of earthaccess you may want to serialize the search results so that can start from there again. Running feature tracking workflows, or InSAR workflows requires generating list of suitbale pairs, and that workflow can get complicated to get to the end "search results" you actually care about, so often requires iteration on methods that use the search results.

At it's simplest this could just be saving the all the `umm_json` results to disk with the search arguments as well. Although, I'd in particular like to be able to save a GeoJSON and/or (good) STAC representation of search results (`asf_search` does this and I pretty much always use that as part of my workflows), but that'd require a umm_json to GeoSON/STAC method, which is a different, but similar feature request (and is effectively being asked for in https://github.com/nsidc/earthaccess/issues/305)

@Sherwin-14, I agree that if we're going to provide serialization methods, it would make sense to provide the Results class first. I have been completely pressed for time, but I could help review that PR if you revive it. And maybe help move it forward in the new year.

**Comment by mfisher87** [comment]
Something I think is important: In addition to serializing the results, we should serialize the original parameters that got those results, and the resulting CMR query. E.g.:

```json
{
  "meta": {
    "export_version": 1,
    "earthaccess_version": "0.19.2"
  },
  "search": {
    "parameters": {
      "collection": ...,
      "temporal": ...,
      "spatial": ...
    },
    "query": {
      "url": "https://cmr.earthdata.nasa.gov/...",
	  "submitted_time": "2025-11-17T18:00:11+00:00"
    }
  },
  "results": [
    ...
  ]
}
```

**Comment by mfisher87** [comment]
Ooo, and it would be really nice to also encode the timestamp the query was executed to get those results. Provenance, baby! Edited example.

**Comment by mfisher87** [comment]
If we're going this route we should have some tests that we're not making breaking changes to the export format.

E.g.:

* With an expected JSON file saved in the repo: 
  * A test that saving results produces an identical JSON to the expected JSON
  * A test that we can load the expected JSON without error and the resulting Results object meets certain conditions (e.g. result count, ??)
* What other ways can we test this?

And we should probably also include the earthaccess version + export version in the JSON. Updated example.

**Comment by itcarroll** [comment]
TLDR: Suggest seeing whether a PR using requests-cache is sufficient, and separate @jhkennedy's proposal into a new issue.

I'm all for something that eliminates repeated calls to CMR when I re-run notebooks. I think there are two visions above: I'll call @mfisher87's a "cache" and @jhkennedy's an "offline catalog" ... maybe that's extreme but my point is to emphasize the difference.

For a cache, I don't support reinventing the wheel and maintaining a JSON schema. The `requests-cache` suggestion looks right on to me. What is it missing?

For an offline catalog, the obvious existing spec is STAC, but this raises a host of challenging questions. Wouldn't this also still benefit from the cache tooling?

**Comment by mfisher87** [comment]
Maybe what would help is writing down our goals and non-goals. For me, these immediately come to mind (I have AGU brain right now so I'll likely have more thoughts later):

Goals:

* Reproducibility; get the same results by using the same data even if CMR has changed
* Provenance; compare current CMR state to CMR state at the time of previous analysis (including exactly when the data was accessed)

**Comment by jhkennedy** [comment]
I'm a bit bogged down this week trying to prep for AGU, but my initial thoughts are:
* I don't trust cached things in general, so I want to know it's happening, preferably needing to explicitly enable it, and I want to be able to easily inspect the cache. That's why I'd prefer an explicit `results.to_*` serialization method, but if requests-cache can be set up that way, 👍 .
* The "offline catalog" is maybe a better description, though I don't know that I'd go so far as to describe it as a "catalog."  I am generally going to be using it outside the earthaccess context, so using the STAC spec makes sense. Really, for me, I  just want a mapping to a GeoDataFrame/geoparquet table. I don't really care if that's STAC-y or just the umm-g flattened, but using STAC enables a lot of other stuff down the line.

**Comment by jhkennedy** [comment]
That said, just serializing the UMM-G JSON objects with some additional search metadata like @mfisher87 proposed would go a _long_ ways towards what I want and I could easily load and flatten that into a (geo)dataframe in just a few lines of code.
